### DISCUSSION TITLE
SD-Card can't be mounted when using LGFX library connected via parallel bus to display

### DISCUSSION THREAD

**Discussion by arturDUS** [lovyan03/LovyanGFX#395] [General]
Hi,
I am using an ESP32 which is connected via 8-bit parallel bus to a display which has the ILI9486 driver implemented.
With the following custom config, the interface and the library is working perfectly:

```
class LGFX : public lgfx::LGFX_Device
{ lgfx::Panel_ILI9486  _panel_instance;   // ILI9486 oder R61581
  lgfx::Bus_Parallel8  _bus_instance;
  public:
    LGFX(void)
    { { auto cfg = _bus_instance.config();
        cfg.i2s_port = I2S_NUM_0;   // 
        cfg.freq_write = 20000000;  // 
        cfg.pin_wr =  16;            // WR 
        cfg.pin_rd =  15;            // RD 
        cfg.pin_rs = 17;            // RS(D/C)
        cfg.pin_d0 = 2;             // D0
        cfg.pin_d1 = 13;            // D1
        cfg.pin_d2 = 26;            // D2
        cfg.pin_d3 = 25;            // D3
        cfg.pin_d4 = 27;            // D4
        cfg.pin_d5 = 12;            // D5
        cfg.pin_d6 = 14;            // D6
        cfg.pin_d7 = 4;             // D7
        _bus_instance.config(cfg);
        _panel_instance.setBus(&_bus_instance);
      }
      { auto cfg = _panel_instance.config();
        cfg.pin_cs           =    33;  // 33
        cfg.pin_rst          =    32;  // 32
        cfg.pin_busy         =    -1;  // -1 = disable
        cfg.panel_width      =   320;  // 
        cfg.panel_height     =   480;  // 
        cfg.offset_x         =     0;  // 
        cfg.offset_y         =     0;  // 
        cfg.offset_rotation  =     2;  // 0..7 (4..7 ist Spiegelverkehrt)
        cfg.dummy_read_pixel =     0;  // 8
        cfg.dummy_read_bits  =     0;  // 1
        cfg.readable         = true;   // 
        cfg.invert           = false;  // 
        cfg.rgb_order        = false;  // 
        cfg.dlen_16bit       = false;  // 
        cfg.bus_shared       = false;  //   false
        _panel_instance.config(cfg);
      }
      setPanel(&_panel_instance);
    }
};
```

If I then try to add a SD card to the hardware using the VSPI bus of the ESP32 and add the code:

```
#include <FS.h> 
#include <SD.h>
SPIClass sdSPI(VSPI);

Setup() {
 
  ....

  sdSPI.begin(SCK, MISO, MOSI, SS);
 
  ....

  if (!SD.begin(5, sdSPI)) {  //   SD.begin(SDchipSelect);
    Serial.println("ERROR: Card Mount Failed");
    SD_Status = 0; // keine SD-Karte
  } else {
    Serial.println("Done");
    SD_Status = 1; // SD-Karte gefunden
  }

  ....

}
```

I alway get the error:

```
Mounting SD-Card...[E][sd_diskio.cpp:769] sdcard_mount(): esp_vfs_fat_register failed 0x(101)
ERROR: Card Mount Failed
```

If I remove the code for the LGFX library from my code, the SD card works and read/write access is possible with the SD card.

So it seems to me, that although I am using a parallel bus config to the display, the SPI interface is also blocked.

Any ideas what I can check?
Thank you for your support.
Artur

**Comment by tobozo** [comment]
hi,

- what happens when only using `SD.begin(5);` alone (no previous SPI.begin)?
- does it improve when psram is disabled?
- what are the values of MISO, MOSI, SCK and SS ? could one of them be conflicting with a tft pin?
- what is the espressif core version, device model and variant? e.g. not all esp32 have the same type/amount of spi buses.

**Reply by arturDUS** [reply]
Commenting out

`sdSPI.begin(SCK, MISO, MOSI, SS);`

and only using

`if (!SD.begin(5)){ ...`

without using the LGFX Library, works for the SD card interface. I can read and write data to the sd-card.
When using both, LGFX library and trying to access the sd-card, sd mount fails. 

... so same behaviour.

output of
```
  Serial.println("Using SPI Bus");
  Serial.print(" SCK: "); Serial.println(SCK);
  Serial.print("MOSI: "); Serial.println(MOSI);
  Serial.print("MISO: "); Serial.println(MISO);
  Serial.print("  SS: "); Serial.println(SS);
  Serial.println();
```
=
```
Using SPI Bus
 SCK: 18
MOSI: 23
MISO: 19
  SS: 5
```
These are the pins that are connected to the sd card.

When programming the ESP it shows
```
Processing wemos_d1_mini32 (platform: espressif32; board: esp32dev; framework: arduino)
------------------------------------------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (3.5.0) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, 
tumpa)
PACKAGES:
 - framework-arduinoespressif32 @ 3.10006.210326 (1.0.6)
 - tool-esptoolpy @ 1.30100.210531 (3.1.0)
 - tool-mkspiffs @ 2.230.0 (2.30)
 - toolchain-xtensa32 @ 2.50200.97 (5.2.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 34 compatible libraries

Chip is ESP32-D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
```

**Comment by tobozo** [comment]
> PLATFORM: Espressif 32 (3.5.0)

Whoops that's based an old espressif core version from early 2022, latest is 6.1.0. 
Your platformio.ini may need to be checked to use the latest versions of core and libraries.

![image](https://user-images.githubusercontent.com/1893754/235303983-fe71cc23-1955-4205-b7e5-82738f603e3e.png)

**Reply by arturDUS** [reply]
Wow, I think I haven't updated the environment for a while... Thanks for the tip.

Unfortunately now I get a lot of core dumps when booting the ESP. I will check that and come back when I could test my code again.

**Reply by tobozo** [reply]
disabling psram may help

**Reply by arturDUS** [reply]
After several tries to get everything running I decidedt to setup platfomio from zero and installed everything (even VS Code) again. Then I created a very basic version of my code and now LGFX and SD are working in parallel.

Thank you for your tips,

Artur
